### DISCUSSION TITLE
Help with constraining an articulated system and body using a spring or soft constraint

### DISCUSSION THREAD

**Discussion by tolstoys19** [sofa-framework/sofa#4642] [Write / Run a simulation] [answered]
Dear Community,

I am working on a project where I need to connect an object to an articulated system. To make it simple right now what I have is a simulation of a cube placed over a joint with 2 links. I want to connect the cube to the links using some type of soft constraint like for example a spring. I know there is the option of using rigidification but as I understand this will completely attach the cube to the link and for us its important for the object attached to be able to slide and move to some degree. I was trying to use a stiff spring to connect both the cube and the links but I have not been able to do it correctly yet and the spring just attaches to the cube but not the link.  Could someone give me some help with this? 

I have uploaded my simulations here.  Please set the mousestiffness to 0.01 or any low value when trying to pull the cube (I still don't know how to set this directly from the code)

[questionconstraint.zip](https://github.com/sofa-framework/sofa/files/14928384/questionconstraint.zip)

Thank you for your time.

**Comment by bakpaul** [comment, marked as answer]
Hi ! 
If you need an attachment allowing some freedom a spring is definitely the right way to do achieve this. 
The problem in your scene is the way you are linking the Spring to the two models. There are two main issues : 
1. Simply putting the spring object in the root node of the object doesn't link the spring to the object. This is done through two datas (one for each object): `object1` and `object2`. What I suggest is, because it belongs to two object and not only one, to put it in the simulation's root node. 
2. This object only works if you try to attach two objects with the same type of DOFS (aka mechanical state's template). Here you were trying to attach a Vec3 object (the cube) to a Vec1 (The articulation), this isn't possible simply because those to objects aren't in the same dimension... You should link to the collision model instead. 

I've added modification to you scene to correct the two aforementioned points here : [DIFFS.zip](https://github.com/sofa-framework/sofa/files/14959941/DIFFS.zip). I have to add that the length data is the rest length of the spring, I have it set to 0 here, and attached your object to the Collision1 model. You could, as you did with the cube, map a single point on the collision of your body to precisely define this attachment point. 

You'll see that I have also added some parameters to use CPU parallelization for some part of the computation to speed up the simulation a bit. I highly encourage you to use a diff tool to look precisely at the modification I did. 

Finally, for the next time, I don't think that I will spend the same time on your simulation if it looks the same as this one... It is very hard to read in such a monolithic block and without any comment. The simulation seems great, readability helps a lot for troubleshooting ;) 

See you with SOFA !!

**Reply by tolstoys19** [reply]
Thank you very much for your quick reply. My question has been answered. I am very sorry about my code being difficult to understand and the lack of comments. I still have to work with SOFA for at least 1 more year so I'll try to be more careful next time.

**Reply by bakpaul** [reply]
It will also ease your process of learning ! By commenting you keep track of your thoughts and logic that you want to acheive ! 

Glad to help you again if needed !
